### PR TITLE
[MM-45296] Fix installation of pre-packaged plugins that are not in the Marketplace

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary

The code in `InstallMarketplacePlugin` did not take into account that a plugin could be prepackaged, but not in the Marketplace. This PR fixes adds an extra check for that case.

I've also broken down the tests in `TestInstallMarketplacePlugin` into sub tests as it was hard to debug individual tests. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-45296
https://mattermost.atlassian.net/browse/MM-44847

#### Release Note
```release-note
Fix installation of pre-packaged plugins that are not in the Marketplace
```

### CODE REVIEW

File: api4/plugin_test.go [hunk 1]
```diff
@@ -1702,34 +1723,58 @@ func TestInstallMarketplacePlugin(t *testing.T) {
 				*cfg.PluginSettings.AllowInsecureDownloadURL = true
 			})
 
-			pRequest = &model.InstallMarketplacePluginRequest{Id: "testplugin2"}
-			manifest2, _, err := client.InstallMarketplacePlugin(pRequest)
-			require.NoError(t, err)
-			require.NotNil(t, manifest2)
-			require.Equal(t, "testplugin2", manifest2.Id)
-			require.Equal(t, "1.2.3", manifest2.Version)
+			t.Run("Install prepackaged, not listed plugin with Marketplace enabled", func(t *testing.T) {
```
Comment (hanzei) [inline api4/plugin_test.go:1726]: It's hard to read in the diff: This is the test added to cover the changes of the PR.

**Comment by hanzei** [comment]
`InstallMarketplacePlugin` is starting to become hard to read. I've tried rewriting it, but the diff became hard to read. Hence, I'm hesitant to include a rewrite into this PR as it makes it harder to understand what the actual fix is.

**Review by levb — DISMISSED** [review]

File: app/plugin_install.go [hunk 2]
```diff
@@ -195,35 +195,38 @@ func (ch *Channels) InstallMarketplacePlugin(request *model.InstallMarketplacePl
 	if *ch.cfgSvc.Config().PluginSettings.EnableRemoteMarketplace {
 		var plugin *model.BaseMarketplacePlugin
 		plugin, appErr = ch.getRemoteMarketplacePlugin(request.Id, request.Version)
-		if appErr != nil {
+		// The plugin might only be prepacked and not on the Marketplace.
+		if appErr != nil && appErr.Id != "app.plugin.marketplace_plugins.not_found.app_error" {
 			return nil, appErr
 		}
```
Comment (lieut-data) [inline app/plugin_install.go:201]: Is the following a potential failure scenario not covered by this change?
* pre-packaged plugin exists
* remote marketplace enabled, but fails to load, triggering an error other than "not found" (e.g. port blocked by firewall)
* installation of pre-packaged plugin still fails

Should we just log a warning here and never return an error from this block? We already return an error below if we fail to find a plugin altogether.

**Review by lieut-data — CHANGES_REQUESTED** [review]
Thanks for tackling this, @hanzei! One thought inline.

File: app/plugin_install.go (hunk 2 above)
Comment (hanzei) [reply app/plugin_install.go]: That is a good suggestion :+1: 

Updated the code in c7b99c8f7

File: app/plugin_install.go [hunk 3]
```diff
@@ -195,35 +195,38 @@ func (ch *Channels) InstallMarketplacePlugin(request *model.InstallMarketplacePl
 	if *ch.cfgSvc.Config().PluginSettings.EnableRemoteMarketplace {
 		var plugin *model.BaseMarketplacePlugin
 		plugin, appErr = ch.getRemoteMarketplacePlugin(request.Id, request.Version)
-		if appErr != nil {
-			return nil, appErr
+		// The plugin might only be prepacked and not on the Marketplace.
```
Comment (lieut-data) [inline app/plugin_install.go:198]:
```suggestion
		// The plugin might only be prepackaged and not on the Marketplace.
```

**Review by lieut-data — APPROVED** [review]

**Comment by hanzei** [comment]
@DHaussermann Heads up that this PR is ready for review

**Comment by DHaussermann** [comment]
@hanzei I've taken another look and still see no difference between this PR and `master`. The plugin can always be installed and enabled. Possibly because it's here https://github.com/mattermost/mattermost-marketplace/blob/master/plugins.json 

> Can you double check that you can install Channel export from the Marketplace once it got disabled?

I'm assuming here that you mean the Channel Export plugin. 

I did follow the steps and regression test the behavior. Could it not be the case that this scenario is no longer applicable with the records currently on the marketplace.

**Review by DHaussermann — APPROVED** [review]
Tested and past
- This PR always worked as expected. But, I struggled to find a case where the install would fail to validate the change
- Worked with Ben to find a case where a server has no licenses file applied that reproduces the problem
- Confirm with this scenario this PR resolves the issue as expected. 

LGTM!

Thanks @hanzei for the help on this one.

**Review by srkgupta — APPROVED** [review]

**Pull request merged**